Relative includes and mappings

jbytheway opened the pull request:
This PR tackles a variety of problems that arise from the interaction of relative includes (by which I mean "" includes which are found relative to the including file) and mappings within IWYU.

Fundamentally, the issue is that quoted includes are used as keys for headers within IWYU.  This works well for system headers, but can break down for local headers used via relative includes. In particular, the IWYU `ConvertToQuotedInclude` method can return different results depending on the including file passed as the second parameter.

The problems manifest in one of two ways:
* Pragmas in headers sometimes appear to have no effect because they were not matched up correctly to a different quoted include for the same header at a different place in the code.
* IWYU suggests `#include`s using the full absolute path to a header, rather than the more logical relative include, because the ultimate includer was not known at the time the quoted include was computed.

The changes here include:
* Add a new visibility map, based on paths rather than quoted includes.  Use it for `IWYU pragma: private` (but not for `IWYU pragma: private, include ...`; see below).
* The targets of mappings (both symbol and include mappings) are now a new `MappedInclude` object rather than simply a quoted include.  The `MappedInclude` contains a quoted include, and sometimes also a path (the path will be used for mappings derived from pragmas, but not mapping files).
* When computing candidate headers, take greater care to avoid includes of absolute paths.
* Use the same canonicalization code for computing candidate headers from both include mappings and symbol mappings; previously they were implemented differently.
* Tests for all the above.

I got a little bit sucked down a rabbit hole with this branch.  But test cases I added to understand one bug kept revealing new bugs, so they are all rather intertwined.  So I think it's reasonable to look at the changes all together.  I claim that much of the code is tidier now than it was; for example there is now less confusion between symbols and quoted includes in some maps.

There are still corner cases I know of that can go wrong; to fix those would require adding a new `IncludeMap` keyed by paths rather than quoted includes.  The lack of that is why I didn't change the behaviour of `IWYU pragma: private, include ...` so the two sorts of private pragma are now inconsistent, which is unfortunate.  However, I think these corner cases are less likely to arise in practice, and I wanted to stop this PR growing any further.

I also happened to notice while working on this that symbol mappings don't respect friend headers properly, but that's definitely a separate issue.

I imagine you'll want the commits squashed or otherwise tidied up before this is merged, but I wanted to get it out for review first.

kimgr commented:
Thank you for this! I've tried to sit down and review it several times since it was posted, but it's quite a large diff. Now I'm on vacation and hopefully I can find some longer unbroken time.

kimgr commented on iwyu_include_picker.cc:
```diff
@@ -1337,13 +1340,22 @@ vector<IncludePicker::MappedInclude>
+  // We also need to consider the header itself.  Make that an option if it's
+  // public or there's no other option.
+  MappedInclude default_header(quoted_header, filepath);
+  if (retval.empty() || GetVisibility(default_header) != kPrivate) {
```
Ah, is this where the map to self thing went? It should default to `kPublic` here too. Oh, wait. `kPublic` is not the same as `kPrivate`, and neither is `kUnusedVisibility`, so I guess this will work as long as the predicates are phrased exactly like this. I think I would feel more comfortable with `kPublic` being spelled out here.

jbytheway replied:
Done.

kimgr commented on iwyu_include_picker.cc:
```diff
@@ -1387,14 +1400,21 @@ vector<string> IncludePicker::GetCandidateHeadersForFilepathIncludedFrom(
-    } else {
+    if (!quoted_include_as_written.empty()) {
       retval.push_back(quoted_include_as_written);
+    } else if (StartsWith(mapped_include.quoted_include, "\"/") &&
```
Is this a check for whether the path is absolute? It probably won't work for Windows paths. I wonder if you should instead strip quotes/angles and use `IsAbsolutePath`...

jbytheway replied:
Done, via a new method on `MappedInclude`.

kimgr commented on iwyu_include_picker.cc:
```diff
@@ -1589,9 +1611,17 @@ IncludeVisibility IncludePicker::ParseVisibility(
 
 IncludeVisibility IncludePicker::GetVisibility(
-    const string& quoted_include) const {
+    const MappedInclude& include) const {
```
Let's add an argument here for the default value:

    GetVisibility(const MappedInclude& include,
                       IncludeVisibility default_value = kUnusedVisibility) const;

... so that callers can override the default behavior where it makes sense to be more explicit.

jbytheway replied:
Done.

kimgr commented on iwyu_include_picker.cc:
```diff
@@ -1589,9 +1611,17 @@ IncludeVisibility IncludePicker::ParseVisibility(
+  if (include_visibility) {
+    return *include_visibility;
+  }
+  if (include.path.empty()) {
```
Is this necessary? I'm guessing we won't have the empty path in the map?

jbytheway replied:
Indeed, tests seem to pass without it.

kimgr commented on iwyu_include_picker.h:
```diff
@@ -81,6 +81,8 @@ class IncludePicker {
   };
   // map_from to <map_to,...>
   typedef map<string, vector<MappedInclude>> IncludeMap;
+  // for tracking header visibility
```
Comments should be full sentences.

jbytheway replied:
Done.

kimgr commented on iwyu_include_picker.h:
```diff
@@ -81,6 +81,8 @@ class IncludePicker {
   };
   // map_from to <map_to,...>
```
Comments should be full sentences.

jbytheway replied:
Done (although this comment was there before I agree it's not a good comment; happy to tidy it up while I'm there).

kimgr commented on iwyu_include_picker.h:
```diff
@@ -102,6 +104,11 @@ class IncludePicker {
 
+  // Indicate that the given path should be considered
+  // a "private" include.  If possible, we use the include-picker
+  // mappings to map such includes to public (not-private) includs.
```
Copy-pasted typo: "includs"

jbytheway replied:
Fixed both.

kimgr commented on iwyu_include_picker.h:
```diff
@@ -194,17 +201,18 @@ class IncludePicker {
   IncludeVisibility ParseVisibility(const string& visibility) const;
 
-  // Return the visibility of a given quoted_include if known, else
+  // Return the visibility of a given mapped_include if known, else
```
`mapped_include` -> `mapped include`

kimgr commented on iwyu_include_picker.h:
```diff
@@ -230,8 +238,15 @@ class IncludePicker {
   // A map of all quoted-includes to whether they're public or private.
-  // Quoted-includes that are not present in this map are assumed public.
-  map<string, IncludeVisibility> filepath_visibility_map_;
+  // Quoted-includes that are not present in this map or the one below
```
This is a little misleading as the path map below does not contain quoted-includes.

How about `Includes that are not present in this map or by path in the one below are assumed public.`

jbytheway replied:
Rephrased in a different way.

kimgr commented on iwyu_include_picker.h:
```diff
@@ -230,8 +238,15 @@ class IncludePicker {
+  VisibilityMap include_visibility_map_;
+
+  // A map of paths to whether they're public or private.
+  // Quoted-includes that are not present in this map or the one above
```
Includes that are not present...

jbytheway replied:
Done.

kimgr commented on iwyu_preprocessor.cc:
```diff
@@ -238,11 +238,10 @@ void IwyuPreprocessorInfo::HandlePragmaComment(SourceRange comment_range) {
+    const string path_this_file = GetFilePath(begin_loc);
+    MutableGlobalIncludePicker()->MarkPathAsPrivate(path_this_file);
+    ERRSYM(this_file_entry) << "Adding private path: "
+                                    << path_this_file << "\n";
```
Indentation

jbytheway replied:
Done.

kimgr commented on run_iwyu_tests.py:
```diff
@@ -183,7 +183,6 @@ def setUp(self):
       'range_for.cc': ['.'],
       're_fwd_decl.cc': ['.'],
       'redecls.cc': ['.'],
-      'relative_include_of_export.cc': ['.'],
```
Now it feels like we lost test coverage of the `-I .` vs no header search path...?

jbytheway replied:
We did.  I think I decided that the version of the test with a `-I` option was mostly redundant.

kimgr commented on iwyu_include_picker.cc:
```diff
@@ -1324,18 +1324,50 @@ string IncludePicker::MaybeGetIncludeNameAsWritten(
   return retval;
 }
 
+vector<IncludePicker::MappedInclude>
```
This wrapping looks nicer to me, but I think it's counter to convention with the return type on its own line. Do you have clang-format installed, and could you use it on the modified regions (not the whole file!)?

jbytheway replied:
Tweaked.  I haven't run clang-format.  Unless you know a clever trick to apply it only to unmodified regions, that seems unwieldy.

kimgr commented on iwyu_include_picker.h:
```diff
@@ -129,7 +129,14 @@ class IncludePicker {
+      const string& symbol) const;
+
+  // As above, but given a specific including header it is possible to convert
+  // MAppedIncludes to quoted include strings (because we can for example know
```
`MAppedIncludes` -> `mapped includes`? Or at least fix the casing :)

jbytheway replied:
Done.

kimgr commented on iwyu_include_picker.h:
```diff
@@ -226,6 +233,11 @@ class IncludePicker {
 
+  // Given a collection of MappedIncludes, and a path that might include them,
+  // choose the best quoted include form for each MappedInclude.
+  vector<string> RelativiseMappedIncludesForPath(
```
`...ForIncluder`?

The method name is a little hard to read, probably because of `Relativise`... I wonder if something like `BestQuotedIncludeForIncluder` would be better? But that's also a mouthful. I'd like to move from "quoted include" to "include name" at some point, but I don't think this is a good time to start.

jbytheway replied:
Went with `BestQuotedIncludesForIncluder`.

kimgr commented on run_iwyu_tests.py:
```diff
@@ -183,6 +185,8 @@ def setUp(self):
       're_fwd_decl.cc': ['.'],
       'redecls.cc': ['.'],
+      'relative_include_of_export_3.cc':
+          ['tests/cxx/relative_include_of_export_3_subfolder'],
```
Hmm. I can't decide whether it's better to have a dedicated sub-directory for each test, or share the existing `subfolder` (soon to be renamed `subdir` in PR #695). In the interest of shorter names, I think I prefer sharing.

jbytheway replied:
Done.

jbytheway commented:
Thanks for the review.  It will take me a while to get back into the head space of this change, but I hope to get a chance to work on it in a week or two.

jbytheway commented on iwyu_include_picker.h:
```diff
@@ -69,7 +69,18 @@ enum IncludeVisibility { kUnusedVisibility, kPublic, kPrivate };
+  // When a symbol or file is mapped to an include, that include is represented
+  // by this struct.  It always has a quoted_include and may also have a path
+  // (depending on its origin).
+  struct MappedInclude {
```
Done.

jbytheway commented on iwyu_include_picker.h:
```diff
@@ -69,7 +69,18 @@ enum IncludeVisibility { kUnusedVisibility, kPublic, kPrivate };
+  // (depending on its origin).
+  struct MappedInclude {
+    explicit MappedInclude(const string& quoted_include,
+                           const string& path = {});
```
Indeed, I'm sceptical too, but one place they work well is for default parameter values.

jbytheway commented on iwyu_output.cc:
```diff
@@ -1241,10 +1241,13 @@ void ProcessFullUse(OneUse* use,
     const vector<string>& method_dfn_files =
-        picker.GetCandidateHeadersForFilepath(GetFilePath(decl_file_entry));
+        picker.GetCandidateHeadersForFilepathIncludedFrom(
+          GetFilePath(decl_file_entry), use_file_path);
```
Done (in fact this part is changed more substantially in my newly refactored version).

jbytheway commented on iwyu_output.cc:
```diff
@@ -1241,10 +1241,13 @@ void ProcessFullUse(OneUse* use,
     const vector<string>& parent_dfn_files =
-        picker.GetCandidateHeadersForFilepath(GetFilePath(parent_file_entry));
+        picker.GetCandidateHeadersForFilepathIncludedFrom(
+          GetFilePath(parent_file_entry), use_file_path);
```
Done.

jbytheway commented on iwyu_output.cc:
```diff
@@ -2066,7 +2069,9 @@ void IwyuFileInfo::HandlePreprocessingDone() {
                       << " as public header for " << private_include
                       << " because used macro is defined by includer.\n";
-        MutableGlobalIncludePicker()->AddMapping(private_include, quoted_file_);
+        IncludePicker::MappedInclude map_to(
```
Indeed.

jbytheway commented on iwyu_preprocessor.cc:
```diff
@@ -847,12 +849,14 @@ void IwyuPreprocessorInfo::PopulateIntendsToProvideMap() {
+    for (const MappedInclude& pub : public_headers_for_header) {
       if (const FileEntry* public_file
-          = GetOrDefault(include_to_fileentry_map_, pub, nullptr)) {
+          = GetOrDefault(include_to_fileentry_map_, pub.quoted_include,
```
Done.

jbytheway commented on tests/cxx/relative_include_of_export-i1.h:
```diff
@@ -0,0 +1,15 @@
+#ifndef INCLUDE_WHAT_YOU_USE_TESTS_CXX_RELATIVE_INCLUDE_OF_EXPORT_I1_H_
+#define INCLUDE_WHAT_YOU_USE_TESTS_CXX_RELATIVE_INCLUDE_OF_EXPORT_I1_H_
+
+enum Nested_Enum {};
```
Done (and similarly elsewhere).

kimgr replied:
Sorry, I gave this a long think when writing up the guidelines, and I came to the conclusion that it's better to name them after the role they play in the test than after the test filename. See docs for rationale.

jbytheway commented on run_iwyu_tests.py:
```diff
@@ -183,6 +183,8 @@ def setUp(self):
       're_fwd_decl.cc': ['.'],
       'redecls.cc': ['.'],
+      'relative_include_of_export.cc': ['.'],
+      'relative_include_of_export_2.cc': [],
```
Agreed.  Now gone.

jbytheway commented on tests/cxx/relative_include_of_export.cc:
```diff
@@ -0,0 +1,21 @@
+//===--- relative_include_of_export-d1.h - test input file for iwyu -------===//
```
I've renamed most of these files now; hopefully all the names are right now.

(Side note: I wish the test files didn't have their own name in them; it makes refactoring a lot more tiresome)

jbytheway commented on tests/cxx/relative_include_of_export_2.cc:
```diff
@@ -0,0 +1,24 @@
+//===--- relative_include_of_export-d1.h - test input file for iwyu -------===//
```
I've renamed all the test files.

jbytheway commented on iwyu_include_picker.cc:
```diff
@@ -875,20 +875,23 @@ bool IsQuotedFilepathPattern(const string& str) {
 // ignoring duplicates.  The input vector is modified in place.
-void ExpandOnce(const IncludePicker::IncludeMap& m, vector<string>* nodes) {
-  vector<string> nodes_and_children;
+void ExpandOnce(const IncludePicker::IncludeMap& m,
```
Done.

jbytheway commented on tests/cxx/relative_include_of_export_3-i2.h:
```diff
@@ -0,0 +1,17 @@
+
+// IWYU pragma: private
+
+enum Nested_Enum {};
```
I went for `EnumI2` to name after the header again.  I don't think duplicated names between tests is a big issue.

jbytheway commented on tests/cxx/relative_include_of_export_3.cc:
```diff
@@ -0,0 +1,21 @@
+//===--- relative_include_of_export-d1.h - test input file for iwyu -------===//
```
For now I've just opted for really long names.  We'll see what you think.

kimgr replied:
Generally, I think they're too long: see guidelines. But I also know it's hard to find short, meaningful names, so I'm not going to veto anything.

I wonder if it would help/be possible to merge all these into a single testcase for the path mapping features of IWYU? Maybe there already are testcases related to this that could be extended?

jbytheway commented on tests/cxx/relative_include_of_export_3-d1.h:
```diff
@@ -0,0 +1,15 @@
+//===--- relative_include_of_export_3-d1.h - test input file for iwyu -----===//
```
Commits have been variously rearranged.

jbytheway commented on iwyu_include_picker.cc:
```diff
@@ -1068,17 +1068,17 @@ void IncludePicker::AddDefaultMappings() {
 }
 
-void IncludePicker::MarkVisibility(const string& quoted_filepath_pattern,
+void IncludePicker::MarkVisibility(VisibilityMap* map,
```
Done.

(Well, I listed the three possible inconsistent sources of quoted include in the commit message; I know at least two of them were being used)

jbytheway commented on iwyu_include_picker.cc:
```diff
@@ -1147,19 +1147,18 @@ void IncludePicker::AddIncludeMapping(const string& map_from,
-  // will self-map them, below).
-  MarkVisibility(map_from, kPrivate);
-  MarkVisibility(map_to.quoted_include, to_visibility);
+  MarkVisibility(&include_visibility_map_, map_to.quoted_include,
```
I think this one is harder to tease apart than the other one.  I haven't tried.  Let me know if you'd still like me to.

kimgr replied:
Actually, I think I found a better way to partition: rename `filepath_visibility_map_` to `include_visibility_map_`, and generalize `MarkVisibility` to take the map as argument.

Then on top of that, introduce `path_visibility_map_`, `MarkPathAsPrivate` and do the fallback in `GetVisibility` (which I think are the only things necessary to fix `pragma: private`?)

That makes it one non-functional change to prepare, and then a much smaller change to fix the bug.

jbytheway replied:
We don't need to mark visibility of `map_from` because it's a symbol, and visibility only matters for files.

jbytheway replied:
I split off that part as you suggested.  I've also separated the refactoring related to the handling of paths being a candidate for themselves that was evidently causing some confusion.

jbytheway commented on iwyu_include_picker.cc:
```diff
@@ -1192,7 +1191,13 @@ void IncludePicker::MarkIncludeAsPrivate(
+  MarkVisibility(&include_visibility_map_, quoted_filepath_pattern, kPrivate);
+}
+
+void IncludePicker::MarkPathAsPrivate(
```
Done.

jbytheway commented on iwyu_include_picker.cc:
```diff
@@ -1301,12 +1307,9 @@ vector<IncludePicker::MappedInclude> IncludePicker::GetPublicValues(
     CHECK_(!StartsWith(value.quoted_include, "@"));
-    if (GetOrDefault(filepath_visibility_map_, value.quoted_include, kPublic)
-        == kPublic)
+    if (GetVisibility(value) != kPrivate)
```
Agreed, that's better.  Done.

jbytheway commented:
OK, I have made a bunch of changes in response to your comments.

The sequence of commits is juggled around a bit.  It turns out I hadn't added the test cases in the same places as the changes that fixed them, which caused me some confusion, but I think I have it straightened out now.  There's one preparatory commit, and then a sequence of further changes together with their respective tests.

There's one more fix here that wasn't there before that I discovered while refactoring the test cases, and seemed like a quick win.

kimgr commented on iwyu_include_picker.cc:
```diff
@@ -1378,18 +1378,19 @@ vector<string> IncludePicker::GetCandidateHeadersForSymbolUsedFrom(
 vector<MappedInclude> IncludePicker::GetCandidateHeadersForFilepath(
     const string& filepath, const string& including_filepath) const {
   CHECK_(has_called_finalize_added_include_lines_ && "Must finalize includes");
+  string absolute_quoted_header = ConvertToQuotedInclude(filepath);
```
In what sense is this absolute?

kimgr replied:
Oh, is `filepath` always absolute, and `ConvertToQuotedInclude` maintains the full path?

jbytheway replied:
It's absolute in the sense that it's not relative to `including_filepath`; it's the "universal" version of this quoted include, which would work for any file in the project.  It won't usually be an absolute path in the usual sense.  I didn't like this choice of name really, but I couldn't think of a better one.  Maybe `universal_quoted_header`?

kimgr commented on iwyu_include_picker.cc:
```diff
@@ -1331,18 +1331,50 @@ string IncludePicker::MaybeGetIncludeNameAsWritten(
-    retval.push_back(mapped_include.quoted_include);
+  for (const MappedInclude& mapped_include : includes) {
+    const string& quoted_include_as_written =
+      MaybeGetIncludeNameAsWritten(including_filepath, mapped_include.path);
```
Continuation indent

jbytheway replied:
Fixed.

jbytheway commented on iwyu_include_picker.h:
```diff
@@ -67,9 +67,17 @@ struct IncludeMapEntry;
+struct MappedInclude {
+  explicit MappedInclude(const string& quoted_include);
+
+  string quoted_include;
```
Generally I am of the opinion that it's unwise to make struct or class fields `const`, because it suppresses the assignment operator, which means you cannot e.g. sort a `vector` of those objects.

kimgr replied:
Ah, I see. Yeah, makes sense. You could probably work around it with some `= delete` and `= default`, but I think that would add unnecessary noise.

jbytheway commented on iwyu_output.cc:
```diff
@@ -2082,7 +2083,8 @@ void IwyuFileInfo::HandlePreprocessingDone() {
                       << " as public header for " << private_include
                       << " because used macro is defined by includer.\n";
-        MutableGlobalIncludePicker()->AddMapping(private_include, quoted_file_);
+        MappedInclude map_to(quoted_file_);
```
Yes, that was a choice for clarity over performance, but happy to go the other way.  The clarity benefit is debatable, anyway.

jbytheway commented on iwyu_preprocessor.cc:
```diff
@@ -228,7 +228,8 @@ void IwyuPreprocessorInfo::HandlePragmaComment(SourceRange comment_range) {
     const string quoted_this_file
         = ConvertToQuotedInclude(GetFilePath(begin_loc));
-    MutableGlobalIncludePicker()->AddMapping(quoted_this_file, suggested);
+    MappedInclude map_to(suggested);
```
Done.

jbytheway commented on iwyu_preprocessor.cc:
```diff
@@ -343,8 +344,9 @@ void IwyuPreprocessorInfo::ProcessHeadernameDirectivesInFile(
     for (string& public_include : public_includes) {
       StripWhiteSpace(&public_include);
       const string quoted_header_name = "<" + public_include + ">";
+      MappedInclude map_to(quoted_header_name);
```
Done.

jbytheway commented on iwyu_preprocessor.cc:
```diff
@@ -411,8 +413,8 @@ void IwyuPreprocessorInfo::MaybeProtectInclude(
         ConvertToQuotedInclude(GetFilePath(includer));
-    MutableGlobalIncludePicker()->AddMapping(include_name_as_written,
-                                             quoted_includer);
+    MappedInclude map_to(quoted_includer);
```
Done.

jbytheway commented on iwyu_include_picker.h:
```diff
@@ -67,10 +67,15 @@ struct IncludeMapEntry;
+                         const string& path = {});
 
   string quoted_include;
+  string path;
```
Same comment applies.

jbytheway commented on tests/cxx/relative_include_of_export.cc:
```diff
@@ -0,0 +1,21 @@
+//
+//===----------------------------------------------------------------------===//
+
+// The purpose of this test is to ensure that the following relative include
```
Done.

jbytheway commented on iwyu_include_picker.cc:
```diff
@@ -1045,6 +1045,15 @@ MappedInclude::MappedInclude(const string& q, const string& p)
 
+bool MappedInclude::HasAbsoluteQuotedInclude() const
+{
+  if (!StartsWith(quoted_include, "\"") || quoted_include.size() < 2) {
```
I agree it's probably cheaper, but it's also practically certain to be false, so putting it first is probably a pessimization.

jbytheway commented on iwyu_include_picker.cc:
```diff
@@ -1045,6 +1045,15 @@ MappedInclude::MappedInclude(const string& q, const string& p)
 }
 
+bool MappedInclude::HasAbsoluteQuotedInclude() const
+{
```
Fixed.

jbytheway commented on iwyu_include_picker.cc:
```diff
@@ -1034,6 +1037,13 @@ string FindFileInSearchPath(const vector<string>& search_path,
 
+MappedInclude::MappedInclude(const string& q)
+  : quoted_include(q)
+{
```
Even with member initializers?  You're not doing that e.g. [here](https://github.com/include-what-you-use/include-what-you-use/blob/93be9c04b6cd913dcd8f5cd1566d116f35208b0f/iwyu_output.cc#L522-L531).

kimgr replied:
I thought so, but if both styles exist I suppose both are right :)

jbytheway commented on iwyu_include_picker.cc:
```diff
@@ -1034,6 +1037,13 @@ string FindFileInSearchPath(const vector<string>& search_path,
+MappedInclude::MappedInclude(const string& q)
+  : quoted_include(q)
+{
+  CHECK_(IsQuotedInclude(quoted_include)
```
Done.

jbytheway commented on iwyu_include_picker.cc:
```diff
@@ -1301,12 +1307,9 @@ vector<IncludePicker::MappedInclude> IncludePicker::GetPublicValues(
   if (!values || values->empty())
     return retval;
 
-  if (GetOrDefault(filepath_visibility_map_, key, kPublic) == kPublic)
```
This was subsumed by the new code in `IncludePicker::GetCandidateHeadersForFilepath` under the "We also need to consider the header itself" comment.  It makes more sense to put it there, because here the key might be a symbol, a full path, or a quoted include.  Only one of those makes sense to add to `retval`.

jbytheway commented:
I believe I've responded to everything except refactoring the tests.

There is some scope for refactoring the tests.  I could use the reusable headers as the target for some of them.  How would you feel about me adding a new reusable header called e.g. `export.h` which includes `indirect.h` with a `pragma: export`?

kimgr commented:
> How would you feel about me adding a new reusable header called e.g. export.h which includes indirect.h with a pragma: export?

Sounds like a good idea to me!

kimgr commented:
The commits come out in a really weird order in the Commits pane. I had to go follow the commit graph to see that the MappedInclude refactoring actually came first :). Could you see if you could touch the dates on the commits to force the order or something?

kimgr commented on iwyu_include_picker.cc:
```diff
@@ -1330,18 +1330,50 @@ string IncludePicker::MaybeGetIncludeNameAsWritten(
   vector<string> retval;
-  for (const MappedInclude& mapped_include : mapped_includes) {
-    retval.push_back(mapped_include.quoted_include);
+  for (const MappedInclude& mapped_include : includes) {
```
Could you add one or more comments to describe the priority order for "best" now? It seems like the quoted-include as written is the least favorite, which comes as a bit of a surprise to me.

I guess that's what the mega-comment above explains, but I wonder if it would be easier to take in as a numbered list or something.

kimgr replied:
Overall, this commit is really nice and focused now, but I can't for the life of me connect the dots to make me understand why it works. :)

jbytheway replied:
I've rewritten the comment.

The as-written version is still the top priority; that's why it's the first alternative in the `if` statement.

jbytheway commented on tests/cxx/relative_include_of_export.cc:
```diff
@@ -0,0 +1,22 @@
+
+/**** IWYU_SUMMARY
+
+(tests/cxx/relative_include_of_export.cc has correct #includes/fwd-decls)
```
For this PR, I have found that it's actually the staying-the-same code paths that are the trickiest ones, so indeed I think this is good.

jbytheway commented on iwyu_preprocessor.cc:
```diff
@@ -411,12 +411,21 @@ void IwyuPreprocessorInfo::MaybeProtectInclude(
+    MappedInclude map_to(quoted_includer, includer_path);
+    MutableGlobalIncludePicker()->AddMapping(include_name_as_written, map_to);
     ERRSYM(includer) << "Adding pragma-export mapping: "
                      << include_name_as_written << " -> " << quoted_includer
```
Done.

jbytheway commented on iwyu_preprocessor.cc:
```diff
@@ -411,12 +411,21 @@ void IwyuPreprocessorInfo::MaybeProtectInclude(
+    if (map_from != include_name_as_written) {
+      MutableGlobalIncludePicker()->AddMapping(map_from, map_to);
+      ERRSYM(includer) << "Adding pragma-export mapping: "
+                       << map_from << " -> " << quoted_includer
```
Done.

jbytheway commented on tests/cxx/relative_include_of_double_export-i2.h:
```diff
@@ -0,0 +1,17 @@
+#ifndef INCLUDE_WHAT_YOU_USE_TESTS_CXX_RELATIVE_INCLUDE_OF_DOUBLE_EXPORT_I2_H_
+#define INCLUDE_WHAT_YOU_USE_TESTS_CXX_RELATIVE_INCLUDE_OF_DOUBLE_EXPORT_I2_H_
+
+// IWYU pragma: private
```
I'm not sure if it's necessary for this test case.  But now I have this test case sharing headers with the one specifically related to exported private headers, so it makes more sense for it to be present.

jbytheway commented on tests/cxx/relative_include_of_double_export-i2.h:
```diff
@@ -0,0 +1,17 @@
+
+// IWYU pragma: private
+
+enum EnumI2 {};
```
These all have new names now.

jbytheway commented:
I think I've addressed everything now.

The tests now share some files in common, and I renamed the most egregiously long one.

I forged the commit dates to get everything in a sensible order.

kimgr commented on iwyu_include_picker.cc:
```diff
@@ -1329,18 +1329,50 @@ string IncludePicker::MaybeGetIncludeNameAsWritten(
+  // 2. If the quoted include in the MappedInclude object is an absolute path,
+  //    that's unlikely to be what's wanted.  Try to convert it to a relative
+  //    include via ConvertToQuotedInclude.
+  // 3. Othewise, use the quoted include present in the MappedInclude.
```
Typo: Othewise

Other than that, this comment is awe-inspiring!

jbytheway replied:
Fixed.

jbytheway commented on tests/cxx/export_private_near.h:
```diff
@@ -0,0 +1,20 @@
+//
+//===----------------------------------------------------------------------===//
+
+// This file is meant to be a generic exporting header.  Simiarly to
```
Fixed this and `export_near.h`.  Think those were the only two with that typo.

jbytheway commented on tests/cxx/relative_include_of_export_added.cc:
```diff
@@ -0,0 +1,29 @@
+// relative include rather than using a full path.
+#include "relative_include_of_export_added-d1.h"
+
+// IWYU: PrivateClass is defined in "export_private_near.h", which isn't directly #included.
```
Done.

kimgr commented:
Thanks! I'm not in any shape to review this from start to finish again, so I'll just make sure it gets merged.

jbytheway commented:
Thanks.
